reference/programming-model.md

```diff
@@ -189,9 +189,102 @@ url := virtualmachine.DnsName().Apply(func(dnsName string) (interface{}, error)
+> Note: The `Output` itself cannot be used directly in string concatenation or other operations, as it is not itself the value of the output.  To transform the value of the output (when it becomes available), the `apply` method should be used instead.  
 
-#### All {#all}
+`.apply` is the most general way to transform one `Output` into another.  However, several common types of transoformations can be done more conveinently.  See (Accessing properties of an Output)[#lifting] and (Working with Outputs and strings)[#ouputs-and-strings] for examples of this.
+
+##### Accessing properties of an Output {#lifting}
+
+It is common to need to only access some subproperty of an `Output` in order to pass in that subproperty to another `Resource`.  For example, when using ACM certificates one might write:
+
+{% include langchoose.html %}
+
+```javascript
+let certCertificate = new aws.acm.Certificate("cert", {
+  domainName: "example.com",
+  validationMethod: "DNS",
+});
+let certValidation = new aws.route53.Record("cert_validation", {
+  // Need to pass along a deep subproperty of this Output
+  records: [certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordValue)],
+  ...
+```
+
+```typescript
+let certCertificate = new aws.acm.Certificate("cert", {
+  domainName: "example.com",
+  validationMethod: "DNS",
+});
+let certValidation = new aws.route53.Record("cert_validation", {
+  // Need to pass along a deep subproperty of this Output
+  records: [certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordValue)],
+  ...
+```
+
+```python
+certificate = aws.acm.Certificate("cert",
+  domain_name="example.com",
+  validation_method="DNS",
+  
+record = aws.route53.Record("validation",
+  # Need to pass along a deep subproperty of this Output
+  records=[certificate.domain_validation_options.apply(
+      lambda domain_validation_options: domain_validation_options[0].resource_record_value
+  )],
+  ...
+```
+
+```go
+// Helpers for accessing properties are not yet available in Go.
+// 
+// See https://github.com/pulumi/pulumi/issues/1614.
+```
+
+To make this kind of property and array-element access more simple, `Outputs` 'lift' the properties of the values they wrap to themselves as well.  For example, if the wrapped property contains a value `prop: string`, then the `Output` wrapping that has a property `prop: Output<string>`.  This 'lifting' is just shorthand for writing out the full `.apply` call.  In other words, an expression of the form `someOutput.someProp` is equivalent to `someOutput.apply(o => o.someProp)`.  Similarly, `someOutput[val]` is equivalent to writing `someOutput.apply(o => o[val])` (with the right syntax for each respective language).  Given this, the above examples can be rewritten into:
```

Comment:
I think we can be more direct here and say “you can access properties of an output directly, and can transform an output with apply”. 

In particular, I don’t think we have to show both the apply and property access versions here - just showing how to access properties seems fine for this section - and noting the limitations that will require using the more general apply.

Reply:
Made more direct.